homeassistant/components/binary_sensor/ihc.py

```diff
@@ -0,0 +1,183 @@
+
+DEPENDENCIES = ['ihc']
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_AUTOSETUP, default='False'): cv.boolean,
+    vol.Optional(CONF_BINARY_SENSORS):
+        [{
+            vol.Required(CONF_ID): cv.positive_int,
+            vol.Optional(CONF_NAME): cv.string,
+            vol.Optional(CONF_TYPE): DEVICE_CLASSES_SCHEMA,
+            vol.Optional(CONF_INVERTING): cv.boolean,
+        }]
+})
+
+
+PRODUCTAUTOSETUP = [
+    # Magnet contact
+    {'xpath': './/product_dataline[@product_identifier="_0x2109"]',
+     'node': 'dataline_input',
+     'type': 'opening',
+     'inverting': True},
+    # Pir sensors
+    {'xpath': './/product_dataline[@product_identifier="_0x210e"]',
+     'node': 'dataline_input',
+     'type': 'motion',
+     'inverting': False},
+    # Pir sensors twilight sensor
+    {'xpath': './/product_dataline[@product_identifier="_0x0"]',
+     'node': 'dataline_input',
+     'type': 'motion',
+     'inverting': False},
+    # Pir sensors alarm
+    {'xpath': './/product_dataline[@product_identifier="_0x210f"]',
+     'node': 'dataline_input',
+     'type': 'motion',
+     'inverting': False},
+    # Smoke detector
+    {'xpath': './/product_dataline[@product_identifier="_0x210a"]',
+     'node': 'dataline_input',
+     'type': 'smoke',
+     'inverting': False},
+    # leak detector
+    {'xpath': './/product_dataline[@product_identifier="_0x210c"]',
+     'node': 'dataline_input',
+     'type': 'moisture',
+     'inverting': False},
+    # light detector
+    {'xpath': './/product_dataline[@product_identifier="_0x2110"]',
+     'node': 'dataline_input',
+     'type': 'light',
+     'inverting': False},
+]
+
+_LOGGER = logging.getLogger(__name__)
+_IHCBINARYSENSORS = {}
+
+
+# pylint: disable=unused-argument
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the IHC binary setsor platform."""
+    ihcplatform = get_ihc_platform(hass)
+    devices = []
+    if config.get(CONF_AUTOSETUP):
+        auto_setup(ihcplatform, devices)
+
+    binarysensors = config.get(CONF_BINARY_SENSORS)
+    if binarysensors:
+        _LOGGER.info("Adding IHC Binary Sensors")
+        for binarysensor in binarysensors:
+            ihcid = binarysensor[CONF_ID]
+            name = (binarysensor[CONF_NAME]
+                    if CONF_NAME in binarysensor
+                    else "ihc_" + str(ihcid))
+            sensortype = (binarysensor[CONF_TYPE]
+                          if CONF_TYPE in binarysensor
+                          else None)
+            inverting = (binarysensor[CONF_INVERTING]
+                         if CONF_INVERTING in binarysensor
+                         else False)
+            add_sensor(devices, ihcplatform.ihc, int(ihcid), name, sensortype,
+                       True, inverting)
+
+    add_devices(devices)
+    # Start notification after devices has been added
+    for device in devices:
+        device.ihc.add_notify_event(device.get_ihcid(),
```

Comment:
Move the registration of this callback to the coroutine `async_added_to_hass` in the entity class `IHCBinarySensor` or even better the super class `IHCDevice` if possible. `async_added_to_hass` will be called when the entity has been added to home assistant and when the entity has access to `schedule_update_ha_state`. See eg https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/binary_sensor/velbus.py#L57-L61.

Reply:
I have move it to the IHCDevice